tools/oeedger8r/Emitter.ml

```diff
@@ -967,274 +472,861 @@ let warn_size_and_count_params (fd : func_decl) =
+        fd.fname
+    ; sprintf "    %s_args_t* pargs_out = (%s_args_t*)output_buffer;" fd.fname
+        fd.fname
     ; ""
-    ; "#include <stdlib.h>"
-    ; "#include <string.h>"
-    ; "#include <wchar.h>"
+    ; "    size_t input_buffer_offset = 0;"
+    ; "    size_t output_buffer_offset = 0;"
+    ; "    OE_ADD_SIZE(input_buffer_offset, sizeof(*pargs_in));"
+    ; "    OE_ADD_SIZE(output_buffer_offset, sizeof(*pargs_out));"
     ; ""
-    ; "OE_EXTERNC_BEGIN"
+    ; (* Buffer validation *)
+      "    /* Make sure input and output buffers are valid. */"
+    ; "    if (!input_buffer || !output_buffer) {"
+    ; "        _result = OE_INVALID_PARAMETER;"
+    ; "        goto done;"
+    ; "    }"
     ; ""
-    ; "/**** ECALL function wrappers. ****/"
-    ; String.concat "\n" oe_gen_host_ecall_wrappers
+    ; (* Prepare in and in-out parameters *)
+      oe_gen_in_and_inout_setters fd.plist
     ; ""
-    ; "/**** OCALL functions. ****/"
-    ; String.concat "\n" oe_gen_ocall_functions
+    ; (* Prepare out and in-out parameters. The in-out parameter is copied to output buffer. *)
+      oe_gen_out_and_inout_setters fd.plist
     ; ""
-    ; "/**** OCALL function table. ****/"
-    ; String.concat "\n" oe_gen_ocall_table
+    ; (* Call the host function *)
+      "    " ^ String.concat "\n    " (oe_gen_call_user_function fd)
     ; ""
-    ; sprintf "oe_result_t oe_create_%s_enclave(" ec.enclave_name
-    ; "    const char* path,"
-    ; "    oe_enclave_type_t type,"
-    ; "    uint32_t flags,"
-    ; "    const void* config,"
-    ; "    uint32_t config_size,"
-    ; "    oe_enclave_t** enclave)"
-    ; "{"
-    ; "    return oe_create_enclave("
-    ; "               path,"
-    ; "               type,"
-    ; "               flags,"
-    ; "               config,"
-    ; "               config_size,"
-    ; sprintf "               __%s_ocall_function_table," ec.enclave_name
-    ; sprintf "               %d," (List.length ufs)
-    ; "               enclave);"
-    ; "}"
+    ; "    /* Propagate errno back to enclave. */"
+    ; ( if uf.uf_propagate_errno then "    pargs_out->_ocall_errno = errno;"
+      else "    /* Errno propagation not enabled. */" )
     ; ""
-    ; "OE_EXTERNC_END"
+    ; (* Mark call as success *)
+      "    /* Success. */"
+    ; "    _result = OE_OK;"
+    ; "    *output_bytes_written = output_buffer_offset;"
+    ; ""
+    ; "done:"
+    ; "    if (pargs_out && output_buffer_size >= sizeof(*pargs_out))"
+    ; "        pargs_out->_result = _result;"
+    ; "}"
     ; "" ]
   in
-  let ecalls_fname = ec.file_shortnm ^ "_u.c" in
-  let os = open_file ecalls_fname ep.untrusted_dir in
-  fprintf os "%s" (String.concat "\n" content) ;
-  close_out os
-
-(** Generate the Enclave code. *)
-let gen_enclave_code (ec : enclave_content) (ep : edger8r_params) =
+  (* Validate Open Enclave supported EDL features. *)
+  let validate_oe_support (ec : enclave_content) (ep : edger8r_params) =
```

Comment:
Oh, should probably move this function further down maybe. I haven't really looked at the ordering of the functions here yet.